### ISSUE TITLE
asyncio: Calling .run() within a task does not behave as per CPython

### ISSUE THREAD

**Issue by peterhinch** [micropython/micropython#15187]
### Port, board and/or hardware

Unix build

### MicroPython version

MicroPython v1.23.0-preview.72.g4a2e510a8.dirty on 2024-02-04; linux [GCC 11.4.0] version

### Reproduction

```py
import asyncio

async def foo(s):
    while True:
        print(f"Task {s}")
        await asyncio.sleep(1)

async def main():
    asyncio.create_task(foo(1))
    await asyncio.sleep(2)
    asyncio.run(foo(2))  # CPython throws a RuntimeError
    await asyncio.sleep(2)
    asyncio.create_task(foo(1))
    await asyncio.sleep(5)
    print("Done")  # never happens in MP

asyncio.run(main())
```

### Expected behaviour

On CPython issuing `run()` in a running task throws a RuntimeError:
```py
Python 3.10.12 (main, Nov 20 2023, 15:14:05) [GCC 11.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import rats17
Task 1
Task 1
Task 1
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/adminpete/rats17.py", line 17, in <module>
    asyncio.run(main())
  File "/usr/lib/python3.10/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/usr/lib/python3.10/asyncio/base_events.py", line 649, in run_until_complete
    return future.result()
  File "/home/adminpete/rats17.py", line 11, in main
    asyncio.run(foo(2))  # CPython throws a RuntimeError
  File "/usr/lib/python3.10/asyncio/runners.py", line 33, in run
    raise RuntimeError(
RuntimeError: asyncio.run() cannot be called from a running event loop
>>> 
```

### Observed behaviour

The `run()` command appears to work, starting the second `foo` instance. However `main` never terminates.
```py
MicroPython v1.23.0-preview.72.g4a2e510a8.dirty on 2024-02-04; linux [GCC 11.4.0] version
Use Ctrl-D to exit, Ctrl-E for paste mode
>>> import rats17
Task 1
Task 1
Task 1
Task 2
Task 1
Task 2
Task 1
Task 2
Task 1
Task 2
... continues forever
```

### Additional Information

This arose from [this issue](https://github.com/peterhinch/micropython-async/issues/120).

A possible fix to `asyncio.core.py` is
```python
def run(coro):
    if cur_task is None:
        return run_until_complete(create_task(coro))
    else:
        raise RuntimeError("asyncio.run() cannot be called from a running event loop")
```

### Code of Conduct

Yes, I agree

**Comment by andrewleech** [comment]
Thanks for your investigation @peterhinch 

I've just been debugging a segfault failure in some unit tests run on unix port which appears to be related to this.

Some of the async test functions were using `asyncio.run(manager.cleanup())` as part of the teardown; this should be just `await manager.cleanup()` but on the face of it ``asyncio.run` looked like it was working. 

However a few tests in, after this pattern had been run once or twice by previous tests, a segfault is thrown.
In case it's of interest this unix port was build with `CFLAGS_EXTRA="-DMICROPY_PY_USELECT_POSIX=0 -DMICROPY_PY_USELECT=1"` 

![image](https://github.com/user-attachments/assets/ae3ce61c-95cf-47ba-90d4-35947bf1957f)

On older versions of micropython these tests passed, however I've just updated to v1.24 and it started crashing.

Long story short this does appear to be quite unsafe to allow asyncio.run to operate inside an existing loop and I think the `RuntimeError` should be raised as suggested.
